### PR TITLE
App Extensions: Added categories module

### LINKED ISSUE #8142: Share Extension: Modular Site Picker - Categories
screen shot 2017-11-14 at 8 34 09 am

* Allow the user assign category for a post (follow site categories UI in site settings)
* Fetch existing categories for site from server (synchronously) 

Depends on: #8139

### PR DESCRIPTION
**Pull request by bummytime**
categories

## Discussion
This PR adds the categories module to the new app extensions. 

### Picker Activation
The category module cell will not launch the category picker if a site only contains a single category (the disclosure chevron will be hidden as well):

no_select

### Defaults
**_IF_** we are able to retrieve the default category for a given site, it will be pre-selected in the category picker screen and shown on the right side of the category module cell:

default_cat

However, if for some reason we are not able determine the default category (customer is not VPNed/proxied, no permissions to access site settings, etc) then the word 'Default' will be displayed:

no_default_cat

Note that the net effect of leaving this default value and saving/publishing the post is the same — a post is created using the default category.

### Data
The Extensions Core Data model (that lives in the shared container) is updated via lightweight migration. A new, optional `String` field was added: `postCategories` to `PostUploadOperation`.

### Network
Two new network functions were added to `AppExtensionsService` 
* `fetchSettingsForSite(_:onSuccess:onFailure:)` — used to get the default category
* `fetchCategoriesForSite(_:onSuccess:onFailure:)` — fetches all categories for a site

In order to prevent expensive network calls from occurring every time a user taps a new site, category info for each site is temporarily stashed in a `NSCache` located in `ShareExtensionAbstractViewController.swift`. This cache is purged when the customer either pulls-to-refresh on the site picker screen, or when exiting the app extension.

### Analytics
Two new analytic values were added and are being tracked: `wpios_*_extension_categories_opened` and `wpios_*_extension_categories_selected`.

Fixes #8142 

## Testing
Make sure to test both the "Save as Draft" action extension as well as the share extension. Both extensions use the same code, but you can never be too careful. 😄 

Testing ideas:
* Test sharing images as well as web/text. Verify the categories are saved properly.
* Try setting some categories and then switching sites (the categories will clear). Verify the correct categories made it to the server.
* Verify the category tree hierarchy — make sure it matches WPiOS (and the server)
* Make sure the cache is working properly and loading categories for each site. If you pull-to-refresh the sites list or cancel, categories should reload when a new site is selected.
* Test the category screens with a spotty network connection.

@frosty and @kwonye would you mind taking a 👀 ?

/cc @folletto for design

### CODE REVIEW

File: WordPress/WordPressShareExtension/ShareData.swift [hunk 1]
```diff
@@ -34,7 +35,70 @@ class ShareData: NSObject {
...
+
+    // MARK: - Computed Vars
+
+    /// Total number of categories for selected site
+    ///
+    var categoryCountForSelectedSite: Int {
+        return allCategoriesForSelectedSite?.count ?? 0
+    }
+
+    /// Computed (read-only) var that returns a comma-delimted string of selected category names. If
```
Comment (frosty) [inline WordPress/WordPressShareExtension/ShareData.swift]: Typo: `delimted` -> `delimited`

File: WordPress/WordPressShareExtension/ShareExtensionAbstractViewController.swift [hunk 2]
```diff
@@ -18,6 +18,25 @@ class ShareExtensionAbstractViewController: UIViewController, ShareSegueHandler
...
+        cache.removeAllObjects()
+    }
+    internal static func storeCategories(_ categories: [RemotePostCategory], for siteID: NSNumber) {
+        cache.setObject(categories as AnyObject, forKey: "\(siteID)-categories" as AnyObject)
+    }
+    internal static func cachedCategoriesForSite(_ siteID: NSNumber) -> [RemotePostCategory]? {
+        return cache.object(forKey: "\(siteID)-categories" as AnyObject) as? [RemotePostCategory]
+    }
+    internal static func storeDefaultCategoryID(_ defaultCategoryID: NSNumber, for siteID: NSNumber) {
+        cache.setObject(defaultCategoryID as AnyObject, forKey: "\(siteID)-default-category" as AnyObject)
```
Comment (frosty) [inline WordPress/WordPressShareExtension/ShareExtensionAbstractViewController.swift]: I know it's only a few short methods, but might be nice to have a small helper to generate the key string for a given site ID for both categories and default category (to save using the literal and `as AnyObject` in two places)

**Comment by bummytime** [comment]
Thank you for the review @frosty! I will address all of that feedback ☝️

**Comment by bummytime** [comment]
Ok, @frosty I _think_ I addressed everything you mentioned above. Thanks again for the your always-great reviews (and for being my human spell check! 😂)! Let me know if any of the loading issues are still occurring.

File: WordPress/WordPressShareExtension/AppExtensionsService.swift [hunk 3]
```diff
@@ -116,6 +136,23 @@ extension AppExtensionsService {
...
+    /// - Parameters:
+    ///   - siteID: Site ID to fetch tags for
+    ///   - onSuccess: Completion handler executed after a successful fetch
+    ///   - onFailure: The failure handler
+    ///
+    func fetchCategoriesForSite(_ siteID: Int, onSuccess: @escaping ([RemotePostCategory]) -> (), onFailure: @escaping FailureWithErrorBlock) {
+        let remote = TaxonomyServiceRemoteREST(wordPressComRestApi: simpleRestAPI, siteID: NSNumber(value: siteID))
+        remote.getCategoriesWithSuccess({ categories in
+            onSuccess(categories)
+        }) { error in
```
Comment (frosty) [inline WordPress/WordPressShareExtension/AppExtensionsService.swift]: Could we keep the `failure: ` parameter name in here too?

**Comment by bummytime** [comment]
@frosty ready for another look. Let me know if you are still seeing the 2nd issue (tapping sites before categories load) you mentioned above ☝️. If you are, I will open another issue to track it. Thanks much!

**Comment by bummytime** [comment]
Thank you @frosty!!!

**Pull request merged**